### ISSUE TITLE
Accept hyphen in crate name in place of underscore

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/cargo#2775]
Crates.io currently accepts hyphens for crates that use underscores, both in the [web interface](https://crates.io/crates/serde-codegen) and in the [API](https://github.com/rust-lang/crates.io/issues/283).

Cargo does not, but should.

```
error: no matching package named `serde-codegen` found (required by `testing`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: *
```

**Comment by dtolnay** [comment]
For `cargo install`, it should install with whatever you asked (so `cargo install my_crate` should create a binary called `my_crate`, `cargo install my-crate` -> `my-crate`).

**Comment by alexcrichton** [comment]
Cargo is somewhat agnostic between `-` and `_`, but it doesn't consider the two characters equivalent. Crates decide whether they want `-` or `_` to begin with and then they must be referenced through that name, disallowing usage of the other.

I'd personally prefer to not accept both `serde-codegen` and `serde_codegen` as it can be confusing to see two different values which mean the same thing in practice from time to time.

**Comment by dtolnay** [comment]
Makes sense, and I don't have a strong preference myself. I saw [this comment](https://users.rust-lang.org/t/is-it-good-practice-to-call-crates-hello-world-hello-world-or-does-it-not-matter/6114/2?u=dtolnay) from @steveklabnik and figured this would be a step toward tooling that reflects the conventions we would like developers to use.

Is there a rationale for crates.io vs cargo behaving differently from each other?

> I'd personally prefer to not accept both `serde-codegen` and `serde_codegen` as it can be confusing to see two different values which mean the same thing in practice from time to time.

Fair, but I would rather see `serde-codegen` and `serde_codegen` and not need to care, vs see `library-a` and `library_b` and need to remember which one to use in each case. "They mean the same thing" is a thing you learn once, while `library-a` and `library_b` is a thing that will bite you for as long as you use Rust.

**Comment by alexcrichton** [comment]
I don't really know why crates.io is agnostic, it wasn't originally and I think that was a patch added after the fact, would have to track that down.

Yeah it's easier to not have to remember, but to me it's more of a downside as it's disguising what's actually happening under the hood.

**Comment by dtolnay** [comment]
The crates.io change was https://github.com/rust-lang/crates.io/commit/89bc5dd46435ec08113c82266ad16464d68d4710.

**Comment by lfairy** [comment]
> I don't really know why crates.io is agnostic, it wasn't originally and I think that was a patch added after the fact, would have to track that down.

The current behavior of crates.io is defined by [RFC 940](https://github.com/rust-lang/rfcs/blob/master/text/0940-hyphens-considered-harmful.md#identify---and-_-on-cratesio):

> Right now, crates.io compares package names case-insensitively. This means, for example, you cannot upload a new package named RUSTC-SERIALIZE because rustc-serialize already exists.
>
> Under this proposal, we will extend this logic to identify - and _ as well.

**Comment by withoutboats** [comment]
I'd like to re-open this issue, I think this is a bug and we should fix it. (Maybe we can talk about it in a cargo meeting).

> I'd personally prefer to not accept both serde-codegen and serde_codegen as it can be confusing to see two different values which mean the same thing in practice from time to time.

At first glance, this makes sense, but I think it doesn't hold as much water when you consider:

* In practice, the name almost always appears in your Cargo.toml exactly once, so its not like they'll mean the same thing in the same file, just that some crates will write them differently.
* If the name contains `-` (which I believe is prefered), both names will already appear, one in the source and one in the manifest, because `-` is not valid in source.

In contrast, users who accidentally add `serde-json` always know what they meant, and because of the conversion users are taught that `-` and `_` are interchangeable in package names. In practice, this seems to be a frustrating wart of cargo which is not adding any clarity for people & possibly even confusing them (if they don't guess that the reason adding `error_chain` failed was that its actually called `error-chain`).

**Comment by dtolnay** [comment]
Reopened. I have also come to feel much more strongly about this in the past year.

**Comment by alexcrichton** [comment]
I also agree this is worthwhile to fix. 

Implementation-wise this won't be easy though, I think, as it'll require changes to the crate index. The changes in Cargo itself after that, though, are likely nominal.

**Comment by withoutboats** [comment]
@alexcrichton That would be a problem, why isn't it sufficient to change how we match the dependency name against the index file to be neutral to underscores (e.g. retry if there's no file with the characters swapped)?

(It'd be better if the index were normalized I agree, but that seems quite challenging).

**Comment by alexcrichton** [comment]
Right now we don't load the entire index in-memory and we currently also don't try to browse the entire index, rather given a crate we drill into exactly which file it's supposed to be. If we have `-` and `_` normalization we'd have a set of filenames that would be the plausible right one, and we'd in theory have to try to check all of them. That's ok on first builds, but we'd need to ensure that if you've got a lock file that this fallback behavior doesn't happen a lot, as it could add up time-wise I think.

**Comment by withoutboats** [comment]
Makes sense, so I think what we should do:

* If the index file for the given ident doesn't exist, fall back by substituting underscores/hyphens in the name
* When generating the lock file, be sure to generate it with the index name, not the name in the toml

Is a divergence between the name in the lock and the toml going to be a problem?

**Comment by alexcrichton** [comment]
Yeah that sounds like it could work!

I think we'll have to maek sure that a `Dependency::name` isn't compared to a `PackageId::name`, although we can perhaps either assume that doesn't happen or otherwise use separate types there if necessary. Sounds plausible at least!

**Comment by Eh2406** [comment]
> If the index file for the given ident doesn't exist, fall back by substituting underscores/hyphens in the name

Is there a better way of doing this then [brute force](https://play.rust-lang.org/?gist=ecf8890098deae4183d541cb9448660b&version=stable&mode=debug&edition=2015)?

**Comment by SimonSapin** [comment]
Brute force take an exponential `O(2^n)` time, but that’s not really a problem when `n` is almost never greater than two.

**Comment by dhardy** [comment]
Why not go further and normalise all names to use hyphens `-`? The first step would be a new version of the query which only returns/finds normalised names; the second step would be a Cargo update to normalise then use the new version. The third step (a bit later perhaps) would be to only show the normalised names on crates.io.

**Comment by Eh2406** [comment]
Because the [index](https://github.com/rust-lang/crates.io-index) wuld need to have both names so that pre-normalise cargo and post-normalise cargo can find it, and that makes for 2 sources of truth.

**Comment by dhardy** [comment]
No it wouldn't if deployed via a new version of Cargo. Unfortunately this would not be backwards compatible (i.e. old versions of Cargo would require correct `-` vs `_`; new versions could accept either).

**Comment by Timmmm** [comment]
> I'd personally prefer to not accept both serde-codegen and serde_codegen as it can be confusing to see two different values which mean the same thing in practice from time to time.

Strongly agree! This is the sort of magic renaming nonsense that CSS has to do because they used `-` everywhere. I think Cargo should issue a warning for new crates that use `-` in the name.

I've actually wasted 10 minutes on this now because `serde-transcode` has a hyphen in it and I haven't worked out how to reference that from Rust code yet. You can't do `use serde-transcode;`. I'm sure there's a workaround but it's an annoying paper cut that I even have to think about it.

If everyone used underscores I wouldn't even have given it a second thought.

**Comment by dhardy** [comment]
@Timmmm agreed that `-` and `_` should not both be allowed and with your logic that `_` is preferable (though I care little). However, any solution has to consider renaming existing crates, which is only viable if `-` and `_` are interchangeable in `Cargo.toml` files *and* the `-` → `_` rename is allowed on `crates.io`.

**Comment by Timmmm** [comment]
Here's what I would do (if I had infinite time etc.). Maybe some of these have been done already - I'm not sure.

1. Add a warning to Cargo for crates with hyphens in the name that new crates with hyphens will not be accepted on crates.io soon.
2. A bit later, stop accepting *new* crates on `crates.io` with hyphens. Also do not accept new underscore crate names that match an existing hyphenated crate.
3. Find crates that only differ in `_`/`-`. Hopefully there aren't any. If there are, resolve by removing/renaming one of them.
4. Accept underscores in `Cargo.toml` for crates that are actually named with hyphens - both for dependencies, and for authors updating existing crates (so they can rename them).
5. Show every crate name on crates.io using underscores (translate the legacy hyphenated ones).

Lots of work though!

**Comment by withoutboats** [comment]
> You can't do `use serde-transcode;`. I'm sure there's a workaround but it's an annoying paper cut that I even have to think about it.

The "workaround" is incredibly simple: you can do `use serde_transcode`. If we don't have diagnostics that recommend this when you type `use serde-transcode` we should.

**Comment by SOF3** [comment]
Since crates.io already does not accept two packages with just `-` and `_` different, why can't we make every place that accepts a package name to silently convert `-` to `_` (or the other way round) internally? It is perfectly feasible that the preference of `-`/`_` of the package author does not affect how users use it (just like whether the package author uses tabs or spaces doesn't affect whether users use tabs or spaces). This can be done with some algorithm similar to how case-insensitive systems handle the names.

BTW, I'm a bit confused. Are we talking about crate names or package names, or are they the same thing?

**Comment by withoutboats** [comment]
> TL;DR: Why do we have to care if ignoring doesn't lead to problems?

That is the goal, and we've solved the problem in both rustc and crates.io but not in cargo, which is why this issue is open.

**Comment by Eh2406** [comment]
> So why is this hard in cargo?

One complication is Alternative Registries, the existing RFC allows registries to have packages that only differ by `-` vs `_`. crates.io does not allow this, but other registries can do what they want. If we want to continue to support the RFC, then Cargo needs to keep the names as is and change all equality checks to equivalency checks. Tracking down every time we use `eq` or `hash` on a type that contains a name and finding a way to make it equivalency... it is going to be hard. (Also a deep well of corner cases.)

Even if we decide to break sutch (nitch) uses of Alternative Registries, we will want a grace period. Some time where it will build with the wrong `-`/`_` but give you a warning that older cargos won't know what you meant. I think this leads to the same implementation problems.

If someone has a way to make this work, I am open to helping make it happen.

**Comment by hskang9** [comment]
@Eh2406  I think I found the problem [here](https://github.com/rust-lang/rust/issues/66676);

**Comment by hskang9** [comment]
@Eh2406 I found the reason why `- `/` _` is all considered as an underscore [here](https://github.com/rust-lang/crates.io/blob/a27c704faa2982ddd75a3dc564da85c0217b950e/migrations/20150319224700_dumped_migration_93/up.sql). The sql function `canon_crate_name` replaces the hyphen with an underscore.

**Comment by carols10cents** [comment]
> crates.io will allow registries to have packages that only differ by `-` vs `_`.

I don't think we want crates.io to allow packages that only differ by `-` vs `_` so I've closed the associated PR; let me know if I've misunderstood.

**Comment by Manishearth** [comment]
I've posted https://internals.rust-lang.org/t/pre-rfc-unify-dashes-and-underscores-on-crates-io/13216/13. It's becoming more clear to me that this doesn't _need_ an RFC, but I've posted the pre RFC anyway.

I don't actually think the exponential growth for names with a large number of separators is a problem. In those cases we can traverse the index trie, looking for both separators whenever there is one, and splitting the search if both exist. Note that this will not cause exponential growth unless there actually are crates with that combination of separator: it's unlikely that `foo-bar-baz-quux-1` and `foo-bar-baz_quux-2` and `foo-bar_baz-quux-3` and so on _all_ exist (and if they do it's probably an automated publish in violation of crates.io policy). In other words, it's only possible to engineer an exponentially bad situation here _on purpose_, it's not really possible as an accident.

**Comment by morkeltry** [comment]
> Strongly agree! This is the sort of magic renaming nonsense that CSS has to do because they used `-` everywhere. I think Cargo should issue a warning for new crates that use `-` in the name.
> 
> I've actually wasted 10 minutes on this now because `serde-transcode` has a hyphen in it and I haven't worked out how to reference that from Rust code yet. You can't do `use serde-transcode;`. I'm sure there's a workaround but it's an annoying paper cut that I even have to think about it.
> 

10 minutes?
I am new to rust and have spent the last two hours working out how to, in my `lib.rs`, import a crate that somebody has named with a hyphen.
I think they named it that way because there's already one with the underscore name and so it is definitely a good way to distinguish between the two versions.

**Comment by SOF3** [comment]
> > Strongly agree! This is the sort of magic renaming nonsense that CSS has to do because they used `-` everywhere. I think Cargo should issue a warning for new crates that use `-` in the name.
> > I've actually wasted 10 minutes on this now because `serde-transcode` has a hyphen in it and I haven't worked out how to reference that from Rust code yet. You can't do `use serde-transcode;`. I'm sure there's a workaround but it's an annoying paper cut that I even have to think about it.
> 
> 10 minutes?
> I am new to rust and have spent the last two hours working out how to, in my `lib.rs`, import a crate that somebody has named with a hyphen.
> I think they named it that way because there's already one with the underscore name and so it is definitely a good way to distinguish between the two versions.

there isn't one with the underscore name. crates.io does not accept an underscore name if there is already one with the hyphen name.

**Comment by LuminousPath** [comment]
If this helps the discussion, I'm a new user to Rust as well and coming from other languages, the ability for a package name and the source code referring to that package name within the same context to differ without being *explicit* is baffling (at least until you google it and end up in a thread like this one).

Moving around in a large project where rust might exist alongside many languages, having to remember that particular quirk about Rust libraries is definitely going to be exhausting in the long term.

**Comment by Pzixel** [comment]
Since there are so many popular crates with `-` in their names (since it's looks nicer and easier to read) isn't it an option to relax naming rules a bit and allow them in rust code in imports? Okay, maybe it's not valid identifier name but it could be a valid crate name.

On the other hand it would forbid it in using in context like `let a = 10 - serde-json::from_str::<i32>("15").unwrap())` so this might look awkward. But people could always manually rename it like `use serde-json as serde_json` if they wish.

**Comment by SOF3** [comment]
Aren't we heading to an overkill here? Perhaps package name conversion to underscore is not something you would realize immediately, but it would be the responsibility of the package developer anyway. Packages can always specify `lib.name` in Cargo.toml to set their own crate name anyway, but how many newbies to rust would actually pay attention to the Cargo.toml of the dependencies they use? I remember getting really confused why `futures-preview` is imported as `futures` instead of `futures_preview` when I just started; if anything, *that* is way more confusing than having `serde-json` imported as `serde_json`, which everyone would immediately understand intuitively (even though they may be doubtful why it is written like that).

I don't know what's the original rationale of allowing underscores, but if we were to start all over, it might be more reasonable to have cargo only take the substring before the first hyphen as the lib crate name (hence the futures-preview case). But now that we are already at this (and it is almost idiomatic to suffix binding crates with `-sys`), collating hyphens as the equivalent of underscore would be most sensible thing to do because that's how everyone understands it.

**Comment by Kimundi** [comment]
Just for reference, Python Package names can also differ in `_`, `-`, and are in the same equivalence class based on a normalization scheme: https://packaging.python.org/en/latest/specifications/name-normalization/#name-normalization

In fact, they take this quite a bit further: Any arbitrary sequence of `_`, `-` or `.` is considered to be the same, `foo.-_-.bar` and `foo-bar` are considered the same. I'm not saying we should do that, just that there is precedence for even more complicated schemes than what is proposed here.

**Comment by Timmmm** [comment]
I would think "Python does this" is pretty strong evidence that it's a terrible idea, given how awful Python's packaging story is!

**Comment by carols10cents** [comment]
> I would think "Python does this" is pretty strong evidence that it's a terrible idea, given how awful Python's packaging story is!

Hey now, we don't language bash here. It's all tradeoffs.

**Comment by epage** [comment]
We discussed this in today's Cargo team meeting.  We are sympathetic to the problem and looked at this from the perspective of how we would solve this.  However, we quickly came to the reluctant conclusion to pass on this when considering how far the experience has improved, reducing the value gained from further improvements which we need to weigh against the amount of policy and coordination work to come up with a solution and roll it out across cargo, crates.io, third-party registries, and third-party tooling that deals with `[dependencies]` tables.

First, let's look at what the experience is like for this.  A user won't have the wrong case in `Cargo.toml` if they start from the TOML snippet on crates.io (provides correct case) or if they run `cargo add` (as it normalizes the case).  Unsure if any LSPs or editor integrations help when directly editing.  If a user does directly add a wrongly-cased dependency to `Cargo.toml`, cargo will suggest the normalized case in the error.

In the future, we could have `cargo fix --broken-code` fix bad casing.  One path to this is combining #13214 with #8283.

Another issue brought up in this thread that is indirectly related to the problem at hand which is translating dependency names to Rust namespaces.  We now have docs.rs and if packages provide examples, that should help clarify things.  For further improvement, I'd recommend creating an issue against rust-lang/rust for trying to detect when `-` is used in a root namespace and suggest changing it to `_`.

As for challenges in resolving this, a relatively new one is that we've also mostly migrated from the git index to the sparse index.  This constrains the solution space because
- We can't enumerate crate names
- Checking if a crate name exists is a network request
- There isn't caching for non-existent crate names

This means that there would be a performance hit on every invocation if we normalize at runtime.  We have dropped the number of case variants we try from 1000 or so to 2 (all `_` and all `-`).

Additional, existing challenges include:
- Support for alternative registries.
  - They are not required to treat crate names that differ by case to be the same (though its a good idea to reduce squatting potential).  Requiring this would be a breaking change that we'd have to approach with caution
  - We are limited in how we evolve the Index and other APIs to not break them (or old cargo)
- Dealing with the design space for a manifest that has two crates in a table that differ by case
- Third-party tooling that works with `[dependencies]` or `[patch]` tables would also need to be updated to handle alternative cases in `Cargo.toml`

**Comment by Timmmm** [comment]
I think there are still some trivial things that could improve matters:

* `cargo init` could warn if the crate name needs normalising
* `cargo add` could print a note saying how to access the crate if the name supplied needs normalising
* crates.io should normalise the name in the `cargo add` command.

Would you accept PRs for those?

**Comment by epage** [comment]
When I said "normalization", I mean from what the user entered to what the canonical name is on crates.io which the crate author controls.

> cargo init could warn if the crate name needs normalising

As this is the crate author controlling it, there would be nothing to normalize.

Whether you mean warning if it doesn't match the Rust namespace for the crate, that would discourage the use of `-` and we are not ready at this time to encourage the ecosystem one way or the other on preferring `-` or `_`

For myself, discussion of the challenge of going from cargo package name to Rust namespace is off topic for this issue and likely needs its own issue.

> cargo add could print a note saying how to access the crate if the name supplied needs normalising

I assume what you mean is to print a note explaining what Rust namespace a package may be accessed by.  This wouldn't just be switching `-` to `_` but also taking into account users overriding `lib.name`.

I think its at least worth exploring (in a separate space as noted above)  My main is balancing is being too helpful and being noisy in our output.  Its not always about one particular message but what principle we apply for preventing "add a message" gradually turning into 10 messages being added.  We've talked about git-like `advice` which can be disabled in config.  It'd help if we had a stable way of setting config on the command-line for easy copy/paste disabling of advice.

> crates.io should normalise the name in the cargo add command.

I'm not sure what this means

**Comment by Timmmm** [comment]
> For myself, discussion of the challenge of going from cargo package name to Rust namespace is off topic for this issue and likely needs its own issue.

Oh sorry, I didn't read closely enough and assumed this was that issue! (I've definitely discussed this before but can't find it now.)

> crates.io should normalise the name in the cargo add command.

I mean [here](https://crates.io/crates/proc-macro2) it says

<img width="290" height="136" alt="Image" src="https://github.com/user-attachments/assets/b7b57af6-ffe0-4058-a395-19abf1fe50ca" />

but it would be more obvious to new users that you access the code via `proc_macro2` if it said `cargo add proc_macro2` (which also works).

I opened a new issue anyway: #15819
